### PR TITLE
builder: Add CreateCountLeadingSignBits

### PR DESCRIPTION
**Pull request by kezhaoAMD**
Add a new CreateCountLeadingSignBits to the builder in LGC that parallels CreateFindSMsb.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

File: lgc/builder/ArithBuilder.cpp [hunk 1]
```diff
@@ -355,6 +355,25 @@ Value *BuilderImpl::aSinACosCommon(Value *x, Constant *coefP0, Constant *coefP1)
...
+  Constant *negOne = ConstantInt::get(value->getType(), -1);
+  Value *result =
+      scalarize(value, [this](Value *value) { return CreateUnaryIntrinsic(Intrinsic::amdgcn_sffbh, value); });
+  if (isFindSMsb)
+    result = CreateSub(ConstantInt::get(value->getType(), 31), result);
+  Value *isNegOne = CreateICmpEQ(value, negOne);
+  Value *isZero = CreateICmpEQ(value, Constant::getNullValue(value->getType()));
+  Value *isNegOneOrZero = CreateOr(isNegOne, isZero);
+  return CreateSelect(isNegOneOrZero, negOne, result, instName);
+}
```
Comment (nhaehnle) [inline lgc/builder/ArithBuilder.cpp:375]: Looking again at the ISA docs, v_ffbh_i32 / v_cls_i32 (which the intrinsic maps to) is already defined as producing -1 if the input is 0 or -1. So:

* In the `!isFindSMsb` case, none of the extra checks are required, the output of amdgcn_sffbh can be used directly
* In the `isFindSMsb` case, I believe that we likely get better codegen on balance by changing the logic to:
```
  Value *isNegOne = CreateICmpEQ(result, negOne);
  result = CreateSub(ConstantInt::get(value->getType(), 31), result);
  return CreateSelect(isNegOne, negOne, result, instName);
```

Overall, I suggest removing the new helper method and instead using the simpler method directly in CreateCountLeadingSignBits, and then implement CreateFindSMsb by calling CreateCountLeadingSignBits.

**Review by nhaehnle — COMMENTED** [review]
Thanks. I did notice another thing after all.

**Comment by nhaehnle** [comment]
ok to test <-- should make amdvlk-admin happy

**Comment by amdvlk-admin** [comment]
### Test summary for commit 467e2b05cb321ffac7f04b3b34a5f444dc5ff62d
<details><summary>CTS tests (Failed: 0/141320)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40065/70667 (56.7%)</li>
<li>Failed: 0/70667 (0.0%)</li>
<li>Not Supported: 30602/70667 (43.3%)</li><li>Warnings: 0/70667 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/ArithBuilder.cpp (hunk 1 above)
Comment (kezhaoAMD) [reply lgc/builder/ArithBuilder.cpp]: Thanks for pointing, code updated.

File: lgc/builder/ArithBuilder.cpp [hunk 2]
```diff
@@ -1310,13 +1310,26 @@ Value *BuilderImpl::CreateFindSMsb(Value *value, const Twine &instName) {
   assert(value->getType()->getScalarType()->isIntegerTy(32));
 
   Constant *negOne = ConstantInt::get(value->getType(), -1);
-  Value *leadingZeroCount =
-      scalarize(value, [this](Value *value) { return CreateUnaryIntrinsic(Intrinsic::amdgcn_sffbh, value); });
+  Value *leadingZeroCount = CreateCountLeadingSignBits(value);
+  Value *isNegOne = CreateICmpEQ(leadingZeroCount, negOne);
```
Comment (nhaehnle) [inline lgc/builder/ArithBuilder.cpp:1314]: The name is misleading: should be leadingSign(Bits)Count. Rest LGTM

Comment (kezhaoAMD) [reply lgc/builder/ArithBuilder.cpp]: updated. Thanks.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 52bcde4011c82b54b210cad362d04940ba9634c7
<details><summary>CTS tests (Failed: 0/141320)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40065/70667 (56.7%)</li>
<li>Failed: 0/70667 (0.0%)</li>
<li>Not Supported: 30602/70667 (43.3%)</li><li>Warnings: 0/70667 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit f04f7febcb3156189c4275f73a5766fdd305ca18
<details><summary>CTS tests (Failed: 0/141320)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40065/70667 (56.7%)</li>
<li>Failed: 0/70667 (0.0%)</li>
<li>Not Supported: 30602/70667 (43.3%)</li><li>Warnings: 0/70667 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]
Thanks, LGTM

**Comment by amdvlk-admin** [comment]
### Test summary for commit 94273d8dac63c17ba2d03d8a06d15e5636f2843a
<details><summary>CTS tests (Failed: 0/141308)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40067/70655 (56.7%)</li>
<li>Failed: 0/70655 (0.0%)</li>
<li>Not Supported: 30588/70655 (43.3%)</li><li>Warnings: 0/70655 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
94273d8dac63c17ba2d03d8a06d15e5636f2843a Jenkins build error.
CMake Error at /home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/CMakeLists.txt:3 (cmake_minimum_required):
  CMake 3.20.0 or higher is required.  You are running version 3.18.0

**Comment by kezhaoAMD** [comment]
> [94273d8](https://github.com/GPUOpen-Drivers/llpc/commit/94273d8dac63c17ba2d03d8a06d15e5636f2843a) Jenkins build error. CMake Error at /home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/CMakeLists.txt:3 (cmake_minimum_required): CMake 3.20.0 or higher is required. You are running version 3.18.0

This error should not related with this PR.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 94273d8dac63c17ba2d03d8a06d15e5636f2843a
<details><summary>CTS tests (Failed: 0/141320)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40065/70667 (56.7%)</li>
<li>Failed: 0/70667 (0.0%)</li>
<li>Not Supported: 30602/70667 (43.3%)</li><li>Warnings: 0/70667 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

**Review by amdrexu — APPROVED** [review]

**Pull request merged**